lib/srv/discovery/fetchers/kube_services.go

```diff
@@ -0,0 +1,335 @@
+	var (
+		appsMu sync.Mutex
+		apps   types.Apps
+	)
+	for _, service := range kubeServices {
+		service := service
+
+		// Skip kubernetes own internal services
+		if isInternalKubeService(service) {
+			continue
+		}
+
+		// Skip service if it has type annotation and it's not 'app'
+		if v, ok := service.GetAnnotations()[types.DiscoveryTypeLabel]; ok && v != services.KubernetesMatchersApp {
+			continue
+		}
+
+		g.Go(func() error {
+			protocolAnnotation := service.GetAnnotations()[types.DiscoveryProtocolLabel]
+
+			ports, err := getServicePorts(service)
+			if err != nil {
+				f.Log.WithError(err).Errorf("could not get ports for the service %q", service.GetName())
+				return nil
+			}
+
+			portProtocols := map[v1.ServicePort]string{}
+			for _, port := range ports {
+				protocol := ""
+				switch protocolAnnotation {
+				case protoHTTPS, protoHTTP, protoTCP:
+					protocol = protocolAnnotation
+				default:
+					protocol = autoProtocolDetection(getServiceFQDN(service), port, f.ProtocolChecker)
+					if protocol == protoTCP {
+						continue
+					}
+				}
+				portProtocols[port] = protocol
+			}
+
+			for port, portProtocol := range portProtocols {
+				if len(portProtocols) == 1 {
+					port.Name = ""
+				}
+				newApp, err := services.NewApplicationFromKubeService(service, f.ClusterName, portProtocol, port)
+				if err != nil {
+					f.Log.WithError(err).Warnf("Could not get app from a Kubernetes service %q, port %d", service.GetName(), port.Port)
+					return nil
+				}
+
+				appsMu.Lock()
+				apps = append(apps, newApp)
+				appsMu.Unlock()
+			}
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return apps.AsResources(), nil
+}
+
+func (f *KubeAppFetcher) ResourceType() string {
+	return types.KindApp
+}
+
+func (f *KubeAppFetcher) Cloud() string {
+	return ""
+}
+
+func (f *KubeAppFetcher) String() string {
+	return fmt.Sprintf("KubeAppFetcher(Namespaces=%v, Labels=%v)", f.Namespaces, f.FilterLabels)
+}
+
+// autoProtocolDetection tries to determine port's protocol. It uses heuristics and port HTTP pinging if needed, provided
+// by protocol checker. It is used when no explicit annotation for port's protocol was provided.
+func autoProtocolDetection(serviceFQDN string, port v1.ServicePort, pc ProtocolChecker) string {
+	if port.AppProtocol != nil {
+		switch p := strings.ToLower(*port.AppProtocol); p {
+		case protoHTTP, protoHTTPS:
+			return p
+		}
+	}
+
+	if port.Port == 443 || strings.EqualFold(port.Name, protoHTTPS) {
+		return protoHTTPS
+	}
+
+	if pc != nil {
+		result := pc.CheckProtocol(fmt.Sprintf("%s:%d", serviceFQDN, port.Port))
+		if result != protoTCP {
+			return result
+		}
+	}
+
+	if port.Port == 80 || port.Port == 8080 || strings.EqualFold(port.Name, protoHTTP) {
+		return protoHTTP
+	}
+
+	return protoTCP
+}
+
+func getServiceFQDN(s v1.Service) string {
+	host := s.GetName()
+	if s.Spec.Type == v1.ServiceTypeExternalName {
```

Comment:
This seems incorrect.

ExternalName services are services that are resolved to external DNS entries instead of services where the targets are simple pods

```go
func getServiceFQDN(s v1.Service) string {
	host := s.GetName()
	return fmt.Sprintf("%s.%s.svc.cluster.local", host, s.GetNamespace())
}

```

Check the example here https://kubernetes.io/docs/concepts/services-networking/service/#externalname

Reply:
Keeping external name check, but returning correctly (after offline discussion)